File: 1 курс/II семестр/антонКурс/dernm.csharp/dernm.cs

using System;

class dernm
{
    /*
    This members must be defined by you:
    static double f(double x)
    */


    /*************************************************************************
    Вычисление M-ой производной путем построения интерполяционного полинома
    по N точкам, размещенным в корнях полинома Чебышева N-ой степени.

    Входные параметры:
        X   -   точка, в которой считаем производную
        H   -   шаг дифференцирования. Точки вычисления функции будут размещены
                на области (X-H, X+H)
        M   -   порядок производной. M>=0
        N   -   число используемых точек, N>=1

    Результат - M-ая производная функции F в точке X
    *************************************************************************/
    public static double derivativenm(double x,
        double h,
        int m,
        int n)
    {
        double result = 0;
        double[] fk = new double[0];
        double[] c = new double[0];
        double[] p = new double[0];
        int i = 0;
        int j = 0;
        int k = 0;
        double b1 = 0;
        double b2 = 0;
        double e = 0;
        double d = 0;

        if( n==0 )
        {
            result = 0;
            return result;
        }
        if( n==1 )
        {
            if( m==0 )
            {
                result = f(x);
            }
            else
            {
                result = 0;
            }
            return result;
        }
        fk = new double[n+1];
        for(k=1; k<=n; k++)
        {
            fk[k] = f(x+h*Math.Cos(Math.PI*(k-0.5)/n));
        }
        c = new double[n-1+1];
        for(j=0; j<=n-1; j++)
        {
            c[j] = 0;
            for(k=1; k<=n; k++)
            {
                c[j] = c[j]+fk[k]*Math.Cos(Math.PI*j*(k-0.5)/n);
            }
            if( j==0 )
            {
                c[j] = c[j]/n;
            }
            else
            {
                c[j] = 2*c[j]/n;
            }
        }
        p = new double[n-1+1];
        for(i=0; i<=n-1; i++)
        {
            p[i] = 0;
        }
        d = 0;
        i = 0;
        do
        {
            k = i;
            do
            {
                e = p[k];
                p[k] = 0;
                if( i<=1 & k==i )
                {
                    p[k] = 1;
                }
                else
                {
                    if( i!=0 )
                    {
                        p[k] = 2*d;
                    }
                    if( k>i+1 )
                    {
                        p[k] = p[k]-p[k-2];
                    }
                }
                d = e;
                k = k+1;
            }
            while( k<=n-1 );
            d = p[i];
            e = 0;
            k = i;
            while( k<=n-1 )
            {
                e = e+p[k]*c[k];
                k = k+2;
            }
            p[i] = e;
            i = i+1;
        }
        while( i<=n-1 );
        if( m>n-1 )
        {
            result = 0;
            return result;
        }
        result = p[m];
        for(i=1; i<=m; i++)
        {
            result = result*i/h;
        }
        return result;
    }
}

File: 1 курс/II семестр/антонКурс/dernm.delphi/dernm.pas
unit dernm;
interface
uses Math, Ap, Sysutils;

(*
This routines must be defined by you:
function F(X : Double):Double;
*)

function DerivativeNM(X : Double; H : Double; M : Integer; N : Integer):Double;

implementation

(*************************************************************************
Вычисление M-ой производной путем построения интерполяционного полинома
по N точкам, размещенным в корнях полинома Чебышева N-ой степени.

Входные параметры:
    X   -   точка, в которой считаем производную
    H   -   шаг дифференцирования. Точки вычисления функции будут размещены
            на области (X-H, X+H)
    M   -   порядок производной. M>=0
    N   -   число используемых точек, N>=1

Результат - M-ая производная функции F в точке X
*************************************************************************)
function DerivativeNM(X : Double; H : Double; M : Integer; N : Integer):Double;
var
    FK : TReal1DArray;
    C : TReal1DArray;
    P : TReal1DArray;
    I : Integer;
    J : Integer;
    K : Integer;
    B1 : Double;
    B2 : Double;
    E : Double;
    D : Double;
begin
    if N=0 then
    begin
        Result := 0;
        Exit;
    end;
    if N=1 then
    begin
        if M=0 then
        begin
            Result := F(X);
        end
        else
        begin
            Result := 0;
        end;
        Exit;
    end;
    SetLength(FK, N+1);
    K:=1;
    while K<=N do
    begin
        FK[K] := F(x+h*Cos(Pi*(K-0.5)/N));
        Inc(K);
    end;
    SetLength(C, N-1+1);
    J:=0;
    while J<=N-1 do
    begin
        C[J] := 0;
        K:=1;
        while K<=N do
        begin
            C[J] := C[J]+FK[K]*Cos(Pi*J*(K-0.5)/N);
            Inc(K);
        end;
        if J=0 then
        begin
            C[J] := C[J]/N;
        end
        else
        begin
            C[J] := 2*C[J]/N;
        end;
        Inc(J);
    end;
    SetLength(P, N-1+1);
    I:=0;
    while I<=N-1 do
    begin
        P[i] := 0;
        Inc(I);
    end;
    d := 0;
    i := 0;
    repeat
        k := i;
        repeat
            e := P[k];
            P[k] := 0;
            if (i<=1) and (k=i) then
            begin
                P[k] := 1;
            end
            else
            begin
                if i<>0 then
                begin
                    P[k] := 2*d;
                end;
                if k>i+1 then
                begin
                    P[k] := P[k]-P[k-2];
                end;
            end;
            d := e;
            k := k+1;
        until  not (k<=N-1);
        d := P[i];
        e := 0;
        k := i;
        while k<=N-1 do
        begin
            e := e+P[k]*C[k];
            k := k+2;
        end;
        P[i] := e;
        i := i+1;
    until  not (i<=N-1);
    if M>N-1 then
    begin
        Result := 0;
        Exit;
    end;
    Result := P[M];
    I:=1;
    while I<=M do
    begin
        Result := Result*I/h;
        Inc(I);
    end;
end;


end.